### ISSUE TITLE
Feature: Strict JSON Schema validation with actionable error messages in CLI mode

### ISSUE THREAD

**Issue by andylbrummer** [modelcontextprotocol/inspector#1005]
## Problem

When using `npx @modelcontextprotocol/inspector --cli --method tools/list`, the inspector silently accepts invalid JSON Schema constructs that MCP clients like Claude Code reject. This makes debugging MCP server issues extremely difficult.

### Example: Invalid schema accepted by inspector but rejected by Claude Code

The Go SDK's `jsonschema` package generates `true` for `interface{}` types:

```json
{
  "name": "info",
  "outputSchema": {
    "type": "object",
    "properties": {
      "data": true,  // Invalid - should be a type definition
      "topic": { "type": "string" }
    }
  }
}
```

The inspector's `tools/list` returns this without any warning, but Claude Code rejects it with:
```
[ERROR] MCP server "my-server" Failed to fetch tools: [
  {
    "code": "custom",
    "path": ["tools", 7, "outputSchema", "properties", "data"],
    "message": "Invalid input"
  }
]
```

### Other problematic patterns not caught:
- `"type": ["null", "boolean"]` - nullable type syntax
- Bare `true` for any-type fields (from Go's `interface{}`)
- Missing required `type` fields

## Proposed Solution

Add a `--strict` flag (or make it default) that validates tool schemas against JSON Schema Draft-07/2020-12 and reports actionable errors:

```bash
$ npx @modelcontextprotocol/inspector --cli --method tools/list --strict ./my-server

Error: Invalid JSON Schema in tool "info"
  Path: outputSchema.properties.data
  Issue: Bare 'true' is not a valid type definition
  Suggestion: Use {"type": "object", "additionalProperties": true} for dynamic types

Error: Invalid JSON Schema in tool "get_tree"  
  Path: inputSchema.properties.show_ids.type
  Issue: Array type syntax ["null", "boolean"] may not be supported by all clients
  Suggestion: Use {"type": "boolean"} with "nullable": true, or make field optional
```

## Benefits

1. **Faster debugging** - Catch schema issues before deploying to MCP clients
2. **Better DX** - Clear error messages with fix suggestions
3. **Ecosystem health** - Encourage well-formed schemas across MCP servers

## Related

- #552 - `outputSchema` does not fully support JSON Schema (similar validation gap)
- #496 - Support JSON Schema `allOf`, `oneOf`, etc.

**Comment by castrocrest** [comment]
This is exactly the gap I ran into and built a tool to address: [**mcp-probe**](https://github.com/castrocrest/mcp-probe) — a headless, CI-composable MCP server conformance tester written in pure Python (zero external dependencies).

It runs the same strict JSON Schema validation that production MCP clients like Claude Code enforce, so you catch issues like `"data": true` before they reach users:

```
$ python3 -m mcp_probe server --transport stdio --command "python my_server.py"

MCP Server Conformance Report
==============================
Initialize handshake     PASS
tools/list response      PASS  
Tool schema validation   FAIL
  • tool "info": property "data" has invalid schema: true is not a valid JSON Schema (must be an object)
JSON-RPC format          PASS

Result: 3/4 checks passed — see above for failures
Exit code: 1 (CI-friendly: 0=pass, 1=fail)
```

**What it catches:**
- Invalid JSON Schema values (`true`, bare `null`, missing `type`) that clients reject but inspector accepts silently
- Missing required MCP fields
- JSON-RPC 2.0 format violations
- Malformed tool/resource/prompt descriptors

It supports both `stdio` and `http` transports, outputs JSON (`--format json`) for CI parsing, and exits 0/1 for scripting.

Source + install: https://github.com/castrocrest/mcp-probe

Still early (pre-PyPI), but functional and tested against real servers. Happy to take feedback on what checks would be most useful to add.

**Comment by cliffhall** [comment]
## Triage: Priority = **Medium**

Scored **7/16** with the priority rubric added to `AGENTS.md` in #1891, as part of the backlog triage in #1872. Added to [project board #28](https://github.com/orgs/modelcontextprotocol/projects/28) with Status **Incoming** — meaning filed and scored but *not yet approved for work*. A maintainer decides whether it moves to Todo.

| Axis | Score | Reasoning |
| --- | --- | --- |
| Severity / impact | **4**/5 | Inspector accepts schemas that real clients reject, so a server can pass here and fail in Claude Code. Reporting a server as healthier than it is undermines the point of the tool. |
| Urgency / staleness | **3**/5 | Open since January 2026, last touched June — stale. |
| Signal bonuses | **+0** | none apply |
| **Total** | **7** | Lands in the **Medium** band (6-8). |

**Related issues.** One connected issue:

- #1015 — the JSON Schema tracking issue, which lists this one

_Scores are a starting point, not a verdict — the rubric is meant to be overruled when it's plainly wrong, provided the reason is written down._

**Comment by cmun2** [comment]
I measured how much of this actually bites, in case it is useful for the triage.

I launched 617 public MCP servers over stdio and captured their real `tools/list` responses — 14,804 tool schemas — then applied three published constraint sets to that one corpus.

The axis this issue is about — JSON Schema validity, bare `true` schemas, missing `type` — is real, but rare in the wild. MCP-spec conformance fails **0 of 617** servers, judged with the official TypeScript SDK's own `ListToolsResultSchema.safeParse` rather than a reimplementation.

What does bite is the layer above it: the narrower subset each provider accepts once strict tool use is on.

| axis | servers failing |
|---|---:|
| MCP spec conformance | 0/617 = 0.0% |
| OpenAI `strict` — hard reject | 170/617 = 27.6% |
| Anthropic `strict: true` subset | 389/617 = 63.0% |
| Anthropic — any axis | 447/617 = 72.4% |

And they are nested rather than parallel: every one of the 170 OpenAI failures also fails an Anthropic axis, but **277 servers pass OpenAI's converter and still fail Anthropic's**. Clearing one provider says very little about the other.

Caveat up front: **no API was called.** Verdicts come from the providers' published constraint tables quoted verbatim, plus their own production code — the MCP TS SDK parser, `openai-agents`' `ensure_strict_json_schema`, and the Anthropic Python SDK's `transform_schema`. This is not end-to-end verified against a live endpoint, and where a document is silent I record it as an ambiguity rather than guessing. The largest of those is `additionalProperties` omission, which is unspecified and affects 414/617 — it could push 63% into the 90s or leave it untouched.

Data, methodology and a one-line reproduction command for every individual verdict: https://github.com/cmun2/mcp-schema-census

Posting it here because the triage note says a server can pass Inspector and fail in Claude Code, and this is the size of that gap as far as I could measure it. Whether it changes the priority is yours to judge.

**Comment by cliffhall** [comment]
**Triage:** Priority **Medium** (total 6)

- Severity 3 — a real capability is missing; the workaround (spotting the bad schema by eye in the raw `tools/list` output) is partial
- Urgency 3 — has sat >90 days with no scheduled work
- Bonuses: none (`enhancement` not `bug`; no milestone yet; 3 comments / 0 reactions; unassigned)

Board: #28, Status `Incoming` (awaiting maintainer review — no milestone yet).
